gravity.sh

```diff
@@ -534,36 +534,65 @@ parseList() {
-
-  local num_target_lines_new num_correct_lines
-  # Get number of lines in source file
-  num_source_lines="$(grep -c "^" "${src}")"
-  # Get the new number of lines in destination file
-  num_target_lines_new="$(grep -c "^" "${target}")"
-  # Number of new correctly added lines
-  num_correct_lines="$(( num_target_lines_new-num_target_lines ))"
-  # Update number of lines in target file
-  num_target_lines="$num_target_lines_new"
-  num_invalid="$(( num_source_lines-num_correct_lines ))"
-  if [[ "${num_invalid}" -eq 0 ]]; then
-    echo "  ${INFO} Analyzed ${num_source_lines} domains"
-  else
-    echo "  ${INFO} Analyzed ${num_source_lines} domains, ${num_invalid} domains invalid!"
-  fi
+  # Remove duplicates from the list
+  mapfile -t unusable_lines <<< "$(sed -r "/([^\.]+\.)+[^\.]{2,}/d" < "${src}")"
+  mapfile -t -O "${#unusable_lines[@]}" unusable_lines <<< "$(sed -r "/[^a-zA-Z0-9.\_-]/!d" < "${src}")"
+  IFS=" " read -r -a unusable_lines <<< "$(tr ' ' '\n' <<< "${unusable_lines[@]}" | sort -u | tr '\n' ' ')"
+
+  # A list of items of common local hostnames not to report as unusable
+  # Some lists (i.e StevenBlack's) contain these as they are supposed to be used as HOST files
+  # but flagging them as unusable causes more confusion than it's worth - so we suppress them from the output
+  false_positives=(
+    "localhost"
+    "localhost.localdomain"
+    "local"
+    "broadcasthost"
+    "localhost"
+    "ip6-localhost"
+    "ip6-loopback"
+    "lo0 localhost"
+    "ip6-localnet"
+    "ip6-mcastprefix"
+    "ip6-allnodes"
+    "ip6-allrouters"
+    "ip6-allhosts"
+    )
+
+  # Read the unusable lines into a string
+  tmp_unusuable_lines_str=" ${unusable_lines[*]} "
+  for false_positive in "${false_positives[@]}"; do
+    # Remove false positives from tmp_unusuable_lines_str
+    tmp_unusuable_lines_str="${tmp_unusuable_lines_str/ ${false_positive} / }"
+  done
+  # Read the string back into an array
+  IFS=" " read -r -a unusable_lines <<< "${tmp_unusuable_lines_str}"
 
-  # Display sample of invalid lines if we found some
-  if [ ${#sample_incorrect_lines[@]} -ne 0 ]; then
-    echo "      Sample of invalid domains:"
-    for each in "${sample_incorrect_lines[@]}"
+  # Get a sample of the incorrect lines, limited to 5 (the list should already have been de-duplicated)
+  IFS=" " read -r -a sample_unusable_lines <<< "$(tr ' ' '\n' <<< "${unusable_lines[@]}" | head -n 5 | tr '\n' ' ')"
+
+  local tmp_new_imported_total
+  # Get the new number of lines in destination file
+  tmp_new_imported_total="$(grep -c "^" "${target}")"
+  # Number of usable lines is the difference between the new total and the old total. (Or, the number of lines we just added.)
+  num_usable_lines="$(( tmp_new_imported_total-num_total_imported_lines ))"
+  # Replace the total with the new total.
+  num_total_imported_lines="$tmp_new_imported_total"
+  # Get the number of unusable lines
+  num_unusable_lines="${#unusable_lines[@]}"
```

Comment:
Maybe this is a clash of semantics vs syntax, but it's not clear what we want to report here: number of unusable lines or unusable domains. How accurate should we be? This number is determined after we removed false-positives and after we removed duplicates.

Image a list has 100 "valid" lines (e.g. no comments) but only 80 domains and 20 non-domains of which 5 are duplicates one one is a false-positive. Which number do you want to show?

Reply:
I'm not sure if we need to be precise about what is not used.
I think the number of imported domains is important and should reflect the final number of imported domains.

To tell truth, the most important feature didn't change: 
Pi-hole is able to filter unusable lines and import real domains found on the list.

Reply:
Report unusable domains.

Reply:
Look at removing the false postives and duplicates as being treated the same as removing comments. We don't include the number of comments in the unsuable total, why would we include the false postives and duplicates?

At the end of the process, either one or two numbers are generated per list. 
 - How many domains we could import (the important one)
 - How many non-domain items were left in the file after stripping out all the comments, duplicates, and false positives.

Reply:
made the variable names even more concise

Reply:
Ok, I see your point. But then we should modify

> How many domains we could import (the important one)

because at the moment this number contains all the duplicated imported domains. We should not de-dup the list before sending it to `${target}` because `FTL` will handel duplicates well. Only before counting.

Reply:
We are really getting into splitting hairs here...

There is no sense in importing duplicate domains from a single list, but a domain existing on several lists is valid - so it makes sense that the would ultimately contain duplicates. (This is reflected in the final total count output by gravity)

How often are domains duplicated In a single list?

Reply:
> How often are domains duplicated In a single list?

I guess it depends on the quality of the list. SB does not contain any duplicates.

I think you convinced me. And with the new wording it's clearer.